### PR TITLE
Use the wheel event instead of DOMMouseScroll

### PR DESCRIPTION
**Pull request by mgol**
(NOTE: I treat it as work in progress. What's the best way to test how this implementation works?)

DOMMouseScroll is a proprietary Firefox event while wheel is specified
in DOM Level 3 Events:
http://www.w3.org/TR/DOM-Level-3-Events/#event-type-wheel

All major browsers implement the wheel event at least in their latest dev
versions (Fx 17, Chrome 31, IE 9, WebKit Nightly) and all older versions of
these browsers except Firefox implement the non-standard mousewheel event.

IE9+ implements both but it doesn't have wheelDeltaX/wheelDeltaY, only
wheelDelta which makes it impossible to detect the scrolling axis. Its wheel
event implements all 3 axises.

The wheel event specifies the deltaMode parameter; 0 means deltas are given
in pixels; 1 means lines and 2 means pages. All browsers seems to specify 0 but
there's no guarantee it doesn't vary by the input device etc. so it's better
to normalize.

wheelDeltaY is usually larger (after passing through abs) than deltaY, in Chrome
that specifies both on both wheel & mousewheel events wheelDeltaY is 3 times
larger.

In this way, all of the browsers have both horizontal & vertical axis recognized
which may be used in the future (for example on OS X you can scroll in both
directions at the same time using a trackpad, creating an effect of panning
via moving 2 fingers around on the trackpad).

### CODE REVIEW

**Comment by fbender** [comment]
@mzgol Would you mind rebasing this PR?
@yurydelendik Are you happy with the general approach now?

**Comment by yurydelendik** [comment]
> Are you happy with the general approach now?

Looks okay. Testing on different platforms (Mac, Win, Linux) and browsers (FF, Chrome, Safari, IE) with mouse and trackpad is required. Please post results/findings about testing.

File: web/viewer.js [hunk 1]
```diff
@@ -1972,21 +1972,32 @@ window.addEventListener('pagechange', function pagechange(evt) {
...
-  var MOUSE_WHEEL_DELTA_FACTOR = 40;
-  var ticks = (evt.type === 'DOMMouseScroll') ? -evt.detail :
-              evt.wheelDelta / MOUSE_WHEEL_DELTA_FACTOR;
-  var direction = (ticks < 0) ? 'zoomOut' : 'zoomIn';
+  var WHEEL_DELTA_FACTOR = 40;
+
+  var delta = evt.deltaY;
+  if (delta !== undefined) {
+    switch (evt.deltaMode) {
+      // case 0: pixels - no change
```
Comment (Snuffleupagus) [inline web/viewer.js]: Won't this cause the `case 0` to be handled in `default` below, and was that really indented here?

Comment (Snuffleupagus) [inline web/viewer.js]: Won't this cause `case 0` to be handled in `default` below, and is that really the intended behaviour here!?

File: web/viewer.js [hunk 2]
```diff
@@ -1972,21 +1972,32 @@ window.addEventListener('pagechange', function pagechange(evt) {
...
+      default: // case 2: pages
+        delta /= document.documentElement.clientHeight;
+    }
+  } else {
+    delta = -evt.wheelDeltaY / 3;
+  }
+  var direction = (delta > 0) ? 'zoomOut' : 'zoomIn';
 
   if (PresentationMode.active) {
-    evt.preventDefault();
```
Comment (Snuffleupagus) [inline web/viewer.js]: Why was this line removed?

**Comment by mzgol** [comment]
Sorry about the delay, I totally forgot about this PR! Rebased.

It still needs to be tested whether the computations have good results; I can try later if you tell me what & how to test.

File: web/viewer.js (hunk 2 above)
Comment (mzgol) [inline web/viewer.js]: Not sure why. It's no longer being removed after the rebase. :)

File: web/viewer.js (hunk 1 above)
Comment (mzgol) [inline web/viewer.js]: Yeah, you're right. Fixed when rebasing.

**Comment by mzgol** [comment]
> Can we polyfill wheel event so we will have only "wheel" event in the main code? Rest of it moved to the web/compatibility.js

I still need to address this comment, don't I? This was what delayed this PR initially, then I didn't have time and later I forgot.

Would this approach I described earlier:
> we can attach a custom event handler to `mousewheel` that will trigger the "custom" `wheel` event

work?

**Comment by fbender** [comment]
> Testing on different platforms (Mac, Win, Linux) and browsers (FF, Chrome, Safari, IE) with mouse and trackpad is required.

Is there a quick and easy way to test this, e.g. a test document I can try in any browser? Are there differences to be expected between website vs. embedded (Fx) rendering/performance that I should be mindful of?

**Comment by brendandahl** [comment]
I've played around with this on OSX and the general behavior of the wheel event:
Touchpad+Chrome:
deltaMode is 0, deltaY is very precise
Touchpad+Firefox:
deltaMode is 0, deltaY is very precise
Wheel+Chrome:
deltaMode is 0, deltaY is incremented at larger values per detent (~4px)
Wheel+Firefox:
deltaMode is 1, deltaY is incremented 1 per detent

The current patch (also the implementation in master) gives an experience that leaves something to be desired. Instead of zooming in and out using zoomIn and zoomOut we should zoom by the actual pixel value that the wheel event gives. I have a rough version of this working, but we need to "debounce" our rendering requests so that we only use CSS transforms to do the scaling and then re-render once the value has settled(maybe 100ms?). This gives a relatively smooth experience, but it's still hard to get a really nice 60fps because of a number of places where we read and write to the DOM which causes multiple reflows.

Fixing it all at once is going to be a lot of work, but we can do it in parts:
1) change page size by actual pixel amount and debounce re-rendering
2) fixup spots where we cause reflows multiple times to get it buttery smooth
3) fixup presentation mode

There's also a whole other piece of work we need to do to make the scrolling in the presentation mode better that is related to this. It's always been kind of terrible, but now that we have actual pixel distances we can improve it. My thought would be to give it more a physical snapping affect. E.g. when the user scrolls we move the page up/down by the pixel distance and once they get to a threshold, say 10px, the page moves up and is changed, if scrolling stops before the page has reach the threshold is snaps back in place.

**Comment by Snuffleupagus** [comment]
Closing as superseded by PR #7181.
